### PR TITLE
🗑️ 💬 Clear chat history - Backend

### LINKED ISSUE #1243: Allow to clear chat history
> > ~I often used chat and there were a lot of messages. Now as soon as I open a chat, the application starts opening all messages from the very first. Opening this message history can take 20-30 seconds, during which I cannot write anything to the chat or even close the tab in the browser. Is it possible to hide the history of the chat as it is made in Facebook messenger?~
> 
> The chat is loaded newest to oldest as a temporary workaround in Talk 7.
The optimal solution is to start at the read marker as described in #1164 
Which will be covered by #1788

Since emptying all chat might still be a useful feature, eg in webinaries before opening the lobby, I will keep this issue open. - @nickvergessen

### PR DESCRIPTION
**Pull request by nickvergessen**
- [x] API to clear the history
- [x] Unshare all files
- [x] Document it
- [x] System message "A has removed the history of the conversation"
- [x] Integration test

Fix #1243

### CODE REVIEW

File: docs/chat.md [hunk 1]
```diff
@@ -153,6 +153,30 @@ See [OCP\RichObjectStrings\Definitions](https://github.com/nextcloud/server/blob
...
+        + `200 OK` - When deleting was successful
+        + `202 Accepted` - When deleting was successful but Matterbridge is enabled so the message was leaked to other services
+        + `403 Forbidden` When the user is not a moderator
+        + `404 Not Found` When the conversation could not be found for the participant
+
+    - Header:
+
+        field | type | Description
+        ---|---|---
+        `X-Chat-Last-Common-Read` | int | ID of the last message read by every user that has read privacy set to public. When the user themself has it set to private the value the header is not set (only available with `chat-read-status` capability)
```
Comment (PVince81) [inline docs/chat.md:173]: I guess this will return the id of the system message that triggers the deletion ?

File: docs/chat.md [hunk 2]
```diff
@@ -268,6 +292,7 @@ See [OCP\RichObjectStrings\Definitions](https://github.com/nextcloud/server/blob
 * `guest_moderator_promoted` - {actor} promoted {user} to moderator
 * `guest_moderator_demoted` - {actor} demoted {user} from moderator
 * `message_deleted` - Message deleted by {actor} (Should not be shown to the user)
+* `cleared_history` - {actor} cleared the history of the conversation
```
Comment (PVince81) [inline docs/chat.md:295]: this key isn't using the same pattern like the others

all the others are "$object_$verb" but here you used "$verb_$object".

</picky_mode>

File: lib/Controller/ChatController.php [hunk 3]
```diff
@@ -592,7 +592,43 @@ public function deleteMessage(int $messageId): DataResponse {
 
 		$bridge = $this->matterbridgeManager->getBridgeOfRoom($this->room);
 
-		$response = new DataResponse($data, $bridge['enabled'] ? Http::STATUS_ACCEPTED: Http::STATUS_OK);
+		$response = new DataResponse($data, $bridge['enabled'] ? Http::STATUS_ACCEPTED : Http::STATUS_OK);
+		if ($this->participant->getAttendee()->getReadPrivacy() === Participant::PRIVACY_PUBLIC) {
```
Comment (PVince81) [inline lib/Controller/ChatController.php:596]: I'm guessing this is an additional fix not directly related to clearing the chat history ?

if yes you might want to make that a separate PR/commit so we can backport, if applicable

**Review by PVince81 — APPROVED** [review]
:+1: some minor comments

File: docs/chat.md (hunk 1 above)
Comment (nickvergessen) [reply docs/chat.md:173]: Actually not, as the chat message is not "read" by everyone, so the value will be 0 basically all the time.

File: docs/chat.md (hunk 2 above)
Comment (nickvergessen) [reply docs/chat.md]: makes sense

File: lib/Controller/ChatController.php (hunk 3 above)
Comment (nickvergessen) [reply lib/Controller/ChatController.php:596]: Yeah we could/should. The header is documented, but the clients are forgiving
I will manually backport it together with the `deleteMessages` fix and the broken integration test

**Review by PVince81 — APPROVED** [review]
:+1:

**Comment by Ivansss** [comment]
Capability `clear-history` is missing :/

**Comment by PVince81** [comment]
> Capability `clear-history` is missing :/

added with https://github.com/nextcloud/spreed/pull/6023
